deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -11,69 +11,118 @@ def upload(options, screenshots)
       return if options[:skip_screenshots]
       return if options[:edit_live]
 
-      app = options[:app]
+      legacy_app = options[:app]
+      app_id = legacy_app.apple_id
+      app = Spaceship::ConnectAPI::App.get(app_id: app_id)
 
-      v = app.edit_version(platform: options[:platform])
-      UI.user_error!("Could not find a version to edit for app '#{app.name}'") unless v
+      platform = Spaceship::ConnectAPI::Platform.map(options[:platform])
+      version = app.get_edit_app_store_version(platform: platform)
+      UI.user_error!("Could not find a version to edit for app '#{app.name}' for '#{platform}'") unless version
+
+      UI.important("Will begin uploading snapshots for '#{version.version_string}' on App Store Connect")
 
       UI.message("Starting with the upload of screenshots...")
       screenshots_per_language = screenshots.group_by(&:language)
 
+      localizations = version.get_app_store_version_localizations
+
       if options[:overwrite_screenshots]
-        UI.message("Removing all previously uploaded screenshots...")
-        # First, clear all previously uploaded screenshots
-        screenshots_per_language.keys.each do |language|
-          # We have to nil check for languages not activated
-          next if v.screenshots[language].nil?
-          v.screenshots[language].each_with_index do |t, index|
-            v.upload_screenshot!(nil, t.sort_order, t.language, t.device_type, t.is_imessage)
+        # Get localizations on version
+        localizations.each do |localization|
+          # Only delete screenshots if trying to upload
+          next unless screenshots_per_language.keys.include?(localization.locale)
+
+          # Iterate over all screenshots for each set and delete
+          screenshot_sets = localization.get_app_screenshot_sets
+          screenshot_sets.each do |screenshot_set|
+            UI.message("Removing all previously uploaded screenshots for '#{localization.locale}' '#{screenshot_set.screenshot_display_type}'...")
+            screenshot_set.app_screenshots.each do |screenshot|
+              UI.verbose("Deleting screenshot - #{localization.locale} #{screenshot_set.screenshot_display_type} #{screenshot.id}")
+              screenshot.delete!
```

Comment:
@joshdholtz this operation takes very long, could we maybe parallelize all those deletion requests, it now makes my builds timeout on Bitrise:
![image](https://user-images.githubusercontent.com/9969006/85507163-270b8180-b5f2-11ea-941f-473e2c9a3dd3.png)

Reply:
I'm trying this approach locally by using the [Parallel](https://github.com/grosser/parallel) gem, so far seems to work fine. It's a drop-in replacement for ruby's native iterators. I've tried changing `upload_screenshots.rb` from this:
```
if options[:overwrite_screenshots]
  # Get localizations on version
  localizations.each do |localization|
```
to this:
```
if options[:overwrite_screenshots]
  # Get localizations on version
  n_threads = [MAX_N_THREADS, localizations.length].min
  Parallel.each(localizations, in_threads: n_threads) do |localization|
```
where `MAX_N_THREADS` is just a constant I've empirically set to 8 to avoid incurring in rate limits.
Similarly for the upload operation, from this:
```
screenshots_per_language.each do |language, screenshots_for_language|
```
to this:
```
n_threads = [MAX_N_THREADS, screenshots_per_language.length].min
Parallel.each(screenshots_per_language, in_threads: n_threads) do |language, screenshots_for_language|
```

It seems to work fine so far, but I've also added a retry mechanism on these operations since the API is a bit flaky.

Reply:
Yeah yeah, I've noticed it takes forever too. Was thinking about async but didn't want to jump the gun on that before other things 😬 But might need to happen.